src/kubernetes.cc

```diff
@@ -1010,6 +1170,72 @@ void KubernetesReader::WatchNodes(
   LOG(INFO) << "Watch thread (node) exiting";
 }
 
+void KubernetesReader::ServiceCallback(
+    MetadataUpdater::UpdateCallback callback,
+    const json::Object* service, Timestamp collected_at, bool is_deleted)
+    throw(json::Exception) {
+  UpdateServiceToMetadataCache(service, is_deleted);
+
+  // TODO: using a temporary did not work here.
+  std::vector<MetadataUpdater::ResourceMetadata> result_vector;
+  result_vector.emplace_back(GetClusterMetadata(collected_at, is_deleted));
+  callback(std::move(result_vector));
+}
+
+void KubernetesReader::WatchServices(MetadataUpdater::UpdateCallback callback) {
+  LOG(INFO) << "Watch thread started for services";
+
+  try {
+    WatchMaster(
+        "Service",
+        std::string(kKubernetesEndpointPath) + "/watch/services/",
+        [=](const json::Object* service, Timestamp collected_at, bool is_deleted) {
+          ServiceCallback(callback, service, collected_at, is_deleted);
+        });
+  } catch (const json::Exception& e) {
+    LOG(ERROR) << e.what();
+    LOG(ERROR) << "No more service metadata will be collected";
+  } catch (const KubernetesReader::QueryException& e) {
+    LOG(ERROR) << "No more service metadata will be collected";
+  }
+  health_checker_->SetUnhealthy("kubernetes_service_thread");
+  LOG(INFO) << "Watch thread (service) exiting";
+}
+
+void KubernetesReader::EndpointsCallback(
+    MetadataUpdater::UpdateCallback callback,
+    const json::Object* endpoints, Timestamp collected_at, bool is_deleted)
+    throw(json::Exception) {
+  UpdateServiceToPodsCache(endpoints, is_deleted);
```

Comment:
What does it mean for an endpoint to be deleted? Can one be deleted without the corresponding service being deleted, or are they always correlated?

Reply:
the two are correlated - endpoints is deleted when the service is deleted.